akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -512,6 +520,20 @@ private[stream] class ActorGraphInterpreter(_initial: GraphInterpreterShell) ext
         false
     }
 
+  //this limits number of messages that can be processed synchronously during one actor receive.
+  private val shortCircuitLimit: Int = _initial.mat.settings.syncProcessingLimit
```

Comment:
The thing is that most pipelines need some external event eventually, so even if you set the limit to Int.MaxValue it will still stop after a few rounds since buffers might become empty. Where it matters is where internal stages chat a log among each other, like a mapConcat+filter, or when a larger graph is created like Http. I used a heuristic here before, but that had no upper cap (which is likely necessary). Maybe a constant is OK, but I would benchmark HTTP and set a constant which has no visible performance impact.